Source/DafnyLanguageServer/CounterExampleGeneration/PartialState.cs

```diff
@@ -0,0 +1,331 @@
+// Copyright by the contributors to the Dafny Project
+// SPDX-License-Identifier: MIT
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.RegularExpressions;
+using Microsoft.Boogie;
+
+namespace Microsoft.Dafny.LanguageServer.CounterExampleGeneration;
+
+public class PartialState {
+
+  public bool IsLoopEntryState => FullStateName.Contains("after some loop iterations");
+  // ghost variables introduced by the counterexample whose values must be true for the counterexample to hold:
+  public List<string> LoopGuards = new();
+  public readonly Dictionary<PartialValue, List<string>> KnownVariableNames = new();
+  private readonly List<PartialValue> initialPartialValues;
+  internal readonly DafnyModel Model;
+  internal readonly Model.CapturedState State;
+  private const string InitialStateName = "<initial>";
+  private static readonly Regex StatePositionRegex = new(
+    @".*\((?<line>\d+),(?<character>\d+)\)",
+    RegexOptions.IgnoreCase | RegexOptions.Singleline
+  );
+  internal readonly Dictionary<Model.Element, PartialValue> AllPartialValues = new();
+
+  private const string BoundVarPrefix = "boundVar";
+
+  internal PartialState(DafnyModel model, Model.CapturedState state) {
+    Model = model;
+    State = state;
+    initialPartialValues = new List<PartialValue>();
+    SetupBoundVars();
+    SetupVars();
+  }
+
+  /// <summary>
+  /// Start with the union of vars and boundVars and expand the set by adding 
+  /// variables that represent fields of any object in the original set or
+  /// elements of any sequence in the original set, etc. This is done
+  /// recursively in breadth-first order and only up to a certain maximum
+  /// depth.
```

Comment:
I decided to get rid of the `depth` parameter altogether in my most recent commit. In practice, we would always want to call this method with `-1` (i.e. no depth limit), since we want all the information from the counterexample that we can get. I think the original motivation for this argument was to potentially limit the size of the counterexample displayed in the IDE. That doesn't actually help though since it only confuses the user by not displaying all of the information, and I think for that reason the Language Server does not use this argument any more (only the old server for Dafny 2 does).